src/main/java/de/tum/in/www1/artemis/security/localvc/LocalVCFilterService.java

```diff
@@ -0,0 +1,412 @@
+            }
+        }
+    }
+
+    private void authorizeUserForExamExercise(String userName, ProgrammingExercise exercise, boolean forPush, User user) throws LocalVCAuthException, LocalVCForbiddenException {
+
+        ProgrammingExerciseStudentParticipation participation = getStudentParticipation(exercise.getId(), user.getLogin());
+
+        // Check that the student owns the participation.
+        if (participation.getStudent().isEmpty() || !participation.getStudent().get().getLogin().equals(userName)) {
+            throw new LocalVCAuthException();
+        }
+
+        Exam exam = exercise.getExerciseGroup().getExam();
+
+        // Start date of the exam must be in the past.
+        if (exam.getStartDate() != null && ZonedDateTime.now().isBefore(exam.getStartDate())) {
+            throw new LocalVCForbiddenException();
+        }
+
+        if (!forPush) {
+            return;
+        }
+
+        // ---- Additional checks for push request. ----
+
+        // Check that submission is in time.
+        if (!examSubmissionService.isAllowedToSubmitDuringExam(exercise, user, false)) {
+            throw new LocalVCForbiddenException();
+        }
+
+        if (participation.isTestRun()) {
+            return;
+        }
+
+        // Check the submission policy.
+        checkSubmissionPolicy(exercise, participation);
+
+        // Check whether there was plagiarism detected and the user was notified by the instructor.
+        if (plagiarismService.wasUserNotifiedByInstructor(participation.getId(), user.getLogin())) {
+            throw new LocalVCForbiddenException();
+        }
+    }
+
+    private void authorizeUserForCourseExercise(Course course, String userName, User user, ProgrammingExercise exercise, boolean forPush) {
+        ProgrammingExerciseStudentParticipation participation;
+        // Check if the repository belongs to a team.
+        if (exercise.isTeamMode()) {
+            participation = authorizeTeam(course.getId(), userName, user, exercise.getId());
+        }
+        else {
+            // ---- Repository belongs to a single student. ----
+
+            // Check that the username in the repository name corresponds to the username used for Basic Auth.
+            if (!user.getLogin().equals(userName)) {
+                throw new LocalVCAuthException();
+            }
+
+            participation = getStudentParticipation(exercise.getId(), user.getLogin());
+
+            // Check that the student owns the participation.
+            if (!participation.isOwnedBy(user)) {
+                throw new LocalVCAuthException();
+            }
+        }
+
+        // Students can commit code for test runs after the due date.
+        // The results for these submissions will not be rated.
+        if (!forPush || participation.isTestRun()) {
+            return;
+        }
+
+        Optional<ZonedDateTime> dueDate = ExerciseDateService.getDueDate(participation);
```

Comment:
I would suggest using `ZonedDateTime startDate = exercise.getParticipationStartDate` as well

Reply:
**Optional** You could even factor out a method like `isBetweenDueAndStartDate`